<a href="https://colab.research.google.com/github/hyeonjin0831/2025_1_Data_analysis/blob/main/3_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**<3-2 잘못된 데이터 수정하기 >**

## **1. 데이터프레임 정보 요약 확인**
**NaN** : 판다스에서 누락된 값 \
**info() 메서드** : 데이터프레임 정보 요약 & 출력 -> Nan이 아닌 값 개수 확인 가능

In [ ]:
#3-1절 178p ns_book4.csv 파일 사용
# gdown 패키지로 코랩에 다운로드
# quiet = False : 다운로드 진행 상황 화면에 표시
import gdown
gdown.download('https://bit.ly/3GisL6J', 'ns_book4.csv', quiet=False)

Downloading...
From: https://bit.ly/3GisL6J
To: /content/ns_book4.csv
100%|██████████| 55.5M/55.5M [00:00<00:00, 124MB/s]


'ns_book4.csv'

In [ ]:
import pandas as pd
ns_book4 = pd.read_csv('ns_book4.csv', low_memory=False)
ns_book4.head()
# low_memory = False : 파일 전체를 한 번에 메모리에 로드
# low_memory = True : 파일을 여러개 작은 조각으로 나누어 읽어들임

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
1     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
2     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
3     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
4     NaN  NaN  NaN    NaN     1   0.0  2021-03-19

In [ ]:
ns_book4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 384591 entries, 0 to 384590
Data columns (total 13 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   번호       384591 non-null  int64  
 1   도서명      384188 non-null  object 
 2   저자       384393 non-null  object 
 3   출판사      379950 non-null  object 
 4   발행년도     384577 non-null  object 
 5   ISBN     384591 non-null  object 
 6   세트 ISBN  56576 non-null   object 
 7   부가기호     310386 non-null  object 
 8   권        63378 non-null   object 
 9   주제분류번호   364727 non-null  object 
 10  도서권수     384591 non-null  int64  
 11  대출건수     384591 non-null  float64
 12  등록일자     384591 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 38.1+ MB


전체 행 개수 : 384,591개 (두 번째 코드)
전체 열 개수 : 13개
'도서명' 열은 NaN이 아닌 행 개수가 384,188개이므로\
전체 행 개수 중 누락된 값은 384591 - 384188 = 403개이다. \


_float64 : 실수형, int64 : 정수형, object : 문자열 or 혼합형 데이터 타입_

< 메모리 사용량 확인하기 >\
info() 메서드 : 원소 개수, 열 데이터 타입 기반으로 메모리 사용량 추정
memory_usage 매개변수 + 'deep'옵션 : 정확한 멤리 사용량

In [ ]:
ns_book4.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 384591 entries, 0 to 384590
Data columns (total 13 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   번호       384591 non-null  int64  
 1   도서명      384188 non-null  object 
 2   저자       384393 non-null  object 
 3   출판사      379950 non-null  object 
 4   발행년도     384577 non-null  object 
 5   ISBN     384591 non-null  object 
 6   세트 ISBN  56576 non-null   object 
 7   부가기호     310386 non-null  object 
 8   권        63378 non-null   object 
 9   주제분류번호   364727 non-null  object 
 10  도서권수     384591 non-null  int64  
 11  대출건수     384591 non-null  float64
 12  등록일자     384591 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 266.3 MB


##**2. 누락된 값 처리하기**
**i) 누락된 값 개수 확인하기 : isna() 메서드 ** \
**isna() 메서드** : 각 행이 비어있는지 나타내는 불리언 배열 반환 -> **sum() 메서드**호출 \
-> 불리언 배열의 T개수로 비어있는 행 개수 얻음 \
=> NaN 직접 카운트

In [ ]:
ns_book4.isna().sum()

번호              0
도서명           403
저자            198
출판사          4641
발행년도           14
ISBN            0
세트 ISBN    328015
부가기호        74205
권          321213
주제분류번호      19864
도서권수            0
대출건수            0
등록일자            0
dtype: int64

< 누락되지 않은 값을 확인하는 메서드 : **notna() 메서드**>\
notna() 메서드 <-> isna() 메서드

**ii) 누락된 값으로 표시하기:None과 np.nan**\
**None** : 정수 저장하는 열에 None저장 -> 누락된 값으로 인식

In [ ]:
#'도서권수'열에 있는 첫 번째 값 None으로 변경
#-> isna() 메서드로 누락된 값 세기

ns_book4.loc[0, '도서권수'] = None
ns_book4['도서권수'].isna().sum()

np.int64(1)

In [ ]:
# head 메서드로 처음 두 개 행 출력

ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN   NaN   0.0  2021-03-19  
1    NaN   1.0   0.0  2021-03-19

판다스는 NaN을 실수값으로 저장\
: 원래 데이터타입이 int64였던 '도서권수'열이 NaN을 표시하기 위해 float64로 자동으로 바꿈\
-> 1.0 이렇게 실수형으로 출력 \

**astype() 메서드** : 데이터타입 지정, 매개변수를 {열 이름: 데이터타입} 형식의 딕셔너리로 전달

In [ ]:
#'도서권수'열과 '대출건수' 열 데이터 타입을 int 32로 바꾸기

ns_book4.loc[0, '도서권수']=1
ns_book4 = ns_book4.astype({'도서권수':'int32', '대출건수':'int32'})
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

In [ ]:
# 문자열 None에 대해서 알아봄
# '부가기호' 열 첫 번째 행에 None 입력

ns_book4.loc[0, '부가기호'] = None
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN  부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  None  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN   NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

문자열 None은 None으로 출력
Nun으로 출력하기 위해서 넘파이 패키지에 있는 **np.nan**사용해야함

In [ ]:
import numpy as np
ns_book4.loc[0, '부가기호'] = np.nan
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

## **iii) 누락된 값 바꾸기(1): loc, fillna() 메서드**
\
**loc 메서드**
- 누락된 값 -> 원하는 값으로 바꿈
- isna() 메서드로 누락된 값 가리키는 **불리언 배열** 생성 -> loc 메서드에 전달

In [ ]:
 #누락된 값을 찾아 불리언 배열로 반환
 set_isbn_na_rows = ns_book4['세트 ISBN'].isna()

#누락된 값 -> 빈 문자열로 바꿈
 ns_book4.loc[set_isbn_na_rows, '세트 ISBN'] =''
#누락된 값 개수 세기
 ns_book4['세트 ISBN'].isna().sum()

 #열의 Nan을 빈 문자열로 바꿔 놓았기 때문에 누락된 행의 개수는 0임

np.int64(0)

**fillna() 메서드**
- 원하는 값 전달하면 NaN 대체 가능
- 새로운 데이터프레임 반환

In [ ]:
ns_book4['부가기호'].fillna('없음').isna().sum()

np.int64(0)

In [ ]:
# 전체 데이터프레임 반환 : {열 이름 : 바꾸려는 값}
ns_book4.fillna({'부가기호': '없음'}).isna().sum()

번호              0
도서명           403
저자            198
출판사          4641
발행년도           14
ISBN            0
세트 ISBN         0
부가기호            0
권          321213
주제분류번호      19864
도서권수            0
대출건수            0
등록일자            0
dtype: int64

iv) **누락된 값 바꾸기(2): replace() 메서드**\
**replace() 메서드** : NaN을 포함한 어떤 값도 바꿀 수 있음 \

**첫째, 바꾸려는 값이 하나일 때**
- replace(원래 값, 새로운 값)


In [ ]:
ns_book4.replace(np.nan, '없음').isna().sum()

번호         0
도서명        0
저자         0
출판사        0
발행년도       0
ISBN       0
세트 ISBN    0
부가기호       0
권          0
주제분류번호     0
도서권수       0
대출건수       0
등록일자       0
dtype: int64

- **둘째, 바꾸려는 값이 여러 개일 때**
- 리스트 형식 전달
- replace([원래 값1, 원래 값2], [새로운 값1, 새로운 값2])
- replace({원래 값1:새로운값1, 원래 값2: 새로운 값2})

In [ ]:
#NaN -> '없음'으로 바꾸고 '2021'문자열 -> '21'로 바꿈
ns_book4.replace([np.nan, '2021'], ['없음', '21']).head(2)

번호           도서명      저자   출판사 발행년도           ISBN 세트 ISBN 부가기호   권 주제분류번호  \
0   1       인공지능과 흙  김동훈 지음   민음사   21  9788937444319           없음  없음     없음   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무   21  9791190123969           없음  없음     없음   

   도서권수  대출건수        등록일자  
0     1     0  2021-03-19  
1     1     0  2021-03-19

In [ ]:
ns_book4.replace({np.nan: '없음', '2021': '21'}).head(2)

번호           도서명      저자   출판사 발행년도           ISBN 세트 ISBN 부가기호   권 주제분류번호  \
0   1       인공지능과 흙  김동훈 지음   민음사   21  9788937444319           없음  없음     없음   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무   21  9791190123969           없음  없음     없음   

   도서권수  대출건수        등록일자  
0     1     0  2021-03-19  
1     1     0  2021-03-19

**셋째, 열 마다 다른 값으로 바꿀 때**
- 딕셔너리 형식
- replace({열 이름: 원래 값}, 새로운 값)

In [ ]:
# '부가기호' 열의 NaN -> '없음'으로 바꿈
ns_book4.replace({'부가기호': np.nan}, '없음').head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319           없음  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969           없음  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

In [ ]:
# '부가기호' 열의 NaN -> '없음'
#'발행년도'열의 '2021' -> '21'으로 바꿈
ns_book4.replace({'부가기호': {np.nan:'없음'}, '발행년도':{'2021':'21'}}).head(2)

번호           도서명      저자   출판사 발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사   21  9788937444319           없음  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무   21  9791190123969           없음  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

## **3. 정규 표현식**

**정규 표현식 (정규식)**
: 문자열 패턴을 찾아 대체하기 위한 규칙의 모음

In [ ]:
ns_book4.replace({'발행년도': {'2021'}})[100:102]

번호                                        도서명  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략   

                                   저자   출판사  발행년도           ISBN 세트 ISBN 부가기호  \
100  헨리 클라우드, 존 타운센드 (지은이), 김진웅 (옮긴이)  좋은씨앗  2018  9788958743019          NaN   
101        로런스 인그래시아 (지은이), 안기순 (옮긴이)    부키  2021  9788960518483          NaN   

       권 주제분류번호  도서권수  대출건수        등록일자  
100  NaN  234.9     1     1  2021-03-15  
101  NaN  325.1     1     0  2021-03-15

**< replace() 메서드 뒤에 대괄호[]를 써도 되나요? >**
- replace() 메서드는 새로운 데이터프레임 반환함

ns_temp = ns_book4.reaplace({'발행년도': {'2021': '21'}})\
ns_temp[100:102]

In [ ]:
ns_temp = ns_book4.replace({'발행년도': {'2021': '21'}})
ns_temp[100:102]

번호                                        도서명  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략   

                                   저자   출판사  발행년도           ISBN 세트 ISBN 부가기호  \
100  헨리 클라우드, 존 타운센드 (지은이), 김진웅 (옮긴이)  좋은씨앗  2018  9788958743019          NaN   
101        로런스 인그래시아 (지은이), 안기순 (옮긴이)    부키    21  9788960518483          NaN   

       권 주제분류번호  도서권수  대출건수        등록일자  
100  NaN  234.9     1     1  2021-03-15  
101  NaN  325.1     1     0  2021-03-15

## **4. 숫자 찾기: \d**
- 정규 표현식에서 숫자를 나타내는 기호 : **\d**
- 네 자리 연도 : **\d\d\d\d**
- 그룹으로 묶기 -> 괄호 : 뒤에 두자리만 묶기 **\d\d(\d\d)**
- 패턴 안에 있는 그룹 : \1,\2 ... (등장순서 순으로)
- **regex 매개변수** = True : 정규표현식 사용


In [ ]:
#'발행년도' 열의 값을 정규 표현식으로 두 자리 연도로 바꿈
# 패턴 안에 있는 첫 번째 그룹으로 연도를 바꿈
ns_book4.replace({'발행년도': {r'\d\d(\d\d)': r'\1'}}, regex=True)[100:102]

번호                                        도서명  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략   

                                   저자   출판사 발행년도           ISBN 세트 ISBN 부가기호  \
100  헨리 클라우드, 존 타운센드 (지은이), 김진웅 (옮긴이)  좋은씨앗   18  9788958743019          NaN   
101        로런스 인그래시아 (지은이), 안기순 (옮긴이)    부키   21  9788960518483          NaN   

       권 주제분류번호  도서권수  대출건수        등록일자  
100  NaN  234.9     1     1  2021-03-15  
101  NaN  325.1     1     0  2021-03-15

**r 문자** : 정규 표현식을 다른 문자열과 구분하기 위함\
**중괄호** : 정규 표현식 반복될 때 개수 지정 \
ex) \d\d = \d{2} -> 연속된 숫자 두 개\
ex) \d{2}(\d{2}) -> 앞에서 사용한 숫자 네 개

In [ ]:
# \d\d(\d\d) = \d{2}(\d{2})
ns_book4.replace({'발행년도': {r'\d{2}(\d{2})': r'\1'}}, regex=True)[100:102]

번호                                        도서명  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략   

                                   저자   출판사 발행년도           ISBN 세트 ISBN 부가기호  \
100  헨리 클라우드, 존 타운센드 (지은이), 김진웅 (옮긴이)  좋은씨앗   18  9788958743019          NaN   
101        로런스 인그래시아 (지은이), 안기순 (옮긴이)    부키   21  9788960518483          NaN   

       권 주제분류번호  도서권수  대출건수        등록일자  
100  NaN  234.9     1     1  2021-03-15  
101  NaN  325.1     1     0  2021-03-15

## **5. 문장 찾기: 마침표(.)**
- **마침표(.)** : 어떤 문자에도 대응하는 정규 표현식 문자
- **\* 문자** : 0개 이상 반복 표시
- **역슬래시(\)** : 괄호를 일반 문자로 인식하기 위함
- **\s** : 공백문자

In [ ]:
#'(지은이)'와 '(옮긴이)' 문자열 삭제함
ns_book4.replace({'저자': {r'(.*)\s\(지은이\)(.*)\s\(옮긴이\)': r'\1\2'},
                  '발행년도' : {r'\d{2}(\d{2})': r'\1'}}, regex=True) [100:102]

번호                                        도서명                    저자  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판  헨리 클라우드, 존 타운센드, 김진웅   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략        로런스 인그래시아, 안기순   

      출판사 발행년도           ISBN 세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
100  좋은씨앗   18  9788958743019          NaN  NaN  234.9     1     1  2021-03-15  
101    부키   21  9788960518483          NaN  NaN  325.1     1     0  2021-03-15

## **6.잘못된 값 바꾸기**
- **contains() 메서드** : 시리즈나 인덱스에서 문자열 패턴을 표현하고 있는지 검사

In [ ]:
#'발행년도' 열을 int 32로 바꿈 -> error 발생 -> 숫자가 아닌 다른 문자가 들어있음
ns_book4.astype({'발행년도':'int32'})

ValueError: invalid literal for int() with base 10: '1988.': Error while type casting for column '발행년도'

In [ ]:
#'발행년도' 열에  '1988'이 포함된 행의 개수 세기
ns_book4['발행년도'].str.contains('1988').sum()

407

< ns_book4['발행년도'] == '1988'처럼 쓰면 안되나요? > \
- ns_book4['발행년도']=='1988' 은 발행년도가 정확히 1988인 것만 찾음. '1988.' 같은 것은 제외됨
- contains() 메서드는 주어진 문자열이 포함된 모든 행을 찾음

- '\d' : 숫자에 대응하는 정규 표현식
- '**\D**' : 숫자가 아닌 다른 모든 문자에 대응하는 표현
- **na 매개변수** = True : 연도가 누락된 행

In [ ]:
# 누락된 값 = 숫자 이외의 문자라고 간주
# 누락된 값 출력
invalid_number = ns_book4['발행년도'].str.contains('\D', na=True)
#na=Tue : 연도가 누락된 행을 True로 표시
print(invalid_number.sum())
ns_book4[invalid_number].head()
#head() 기본값이 5여서 5개만 출력됨

1777


번호                       도서명  \
19138  19565                      단국강토   
19227  19736                    삼성의 역사   
26097  26812                  배고픈 애벌레    
29817  30586  (The) Sopranos sessions    
29940  30709                    다음엔 너야   

                                                      저자                 출판사  \
19138                                              홍태수 저             매일경제신문사   
19227                                              송부웅 撰                  삼양   
26097                                  에릭 칼 글·그림 ;이희재 옮김               더큰컴퍼니   
29817  Matt Zoller Seitz,$eAlan Sepinwall,$eLaura Lip...  Harry N Abrams Inc   
29940                         에른스트 얀들 글;노르만 융에 그림;박상순 옮김                 비룡소   

               발행년도           ISBN        세트 ISBN 부가기호    권  주제분류번호  도서권수  \
19138         1988.  9788974420031                 NaN  NaN   511.1     1   
19227  단기4334[2001]  9788985464369                   0  NaN  911.02     1   
26097        [2019]  9788959514083                 NaN  NaN     843     1   
29817         2019.  9781419734946                 NaN  NaN  326.76     1   
29940         2018.  9788949110646  9788949110004    7  NaN     853     1   

       대출건수        등록일자  
19138     1  2019-12-19  
19227     1  2019-12-06  
26097     0  2019-08-12  
29817     0  2019-06-13  
29940     9  2019-06-04

In [ ]:
#앞 뒤에 있는 문자 제외시킴
ns_book5 = ns_book4.replace({'발행년도':r'.*(\d{4}).*'}, r'\1', regex=True)
ns_book5[invalid_number].head()

번호                       도서명  \
19138  19565                      단국강토   
19227  19736                    삼성의 역사   
26097  26812                  배고픈 애벌레    
29817  30586  (The) Sopranos sessions    
29940  30709                    다음엔 너야   

                                                      저자                 출판사  \
19138                                              홍태수 저             매일경제신문사   
19227                                              송부웅 撰                  삼양   
26097                                  에릭 칼 글·그림 ;이희재 옮김               더큰컴퍼니   
29817  Matt Zoller Seitz,$eAlan Sepinwall,$eLaura Lip...  Harry N Abrams Inc   
29940                         에른스트 얀들 글;노르만 융에 그림;박상순 옮김                 비룡소   

       발행년도           ISBN        세트 ISBN 부가기호    권  주제분류번호  도서권수  대출건수  \
19138  1988  9788974420031                 NaN  NaN   511.1     1     1   
19227  2001  9788985464369                   0  NaN  911.02     1     1   
26097  2019  9788959514083                 NaN  NaN     843     1     0   
29817  2019  9781419734946                 NaN  NaN  326.76     1     0   
29940  2018  9788949110646  9788949110004    7  NaN     853     1     9   

             등록일자  
19138  2019-12-19  
19227  2019-12-06  
26097  2019-08-12  
29817  2019-06-13  
29940  2019-06-04

In [ ]:
#숫자 이외의 문자만 들어 행의 개수와 데이터 확인
unkown_year = ns_book5['발행년도'].str.contains('\D', na=True)
print(unkown_year.sum())
ns_book5[unkown_year].head()

67


번호                도서명        저자      출판사     발행년도           ISBN  \
30838    31616       본격 한중일 세계사 5  굽시니스트 지음   위즈덤하우스      NaN  9791190065092   
39130    40141     정책금융의 현황과 발전과제   정책금융연구회   한국산업은행      NaN  9788992784108   
39256    40268  서울지역 유적 발굴조사 총서 3   서울역사박물관  서울역사박물관      NaN  9791186324714   
76836    81202            흰머리 큰줄기    한호진 지음    秀文出版社  [발행년불명]  9788973010769   
150543  160436          (속) 경제학사     박장환 지음      NaN   [20--]  9788994339207   

              세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
30838                  NaN  NaN    NaN     0     0  2019-05-28  
39130                  NaN  NaN  327.1     1     0  2019-01-22  
39256   9791186324431  NaN  NaN    NaN     0     0  2019-01-22  
76836                    0  NaN  699.1     1     1  2016-11-10  
150543                   1  NaN  320.9     1     2  2012-11-19

In [ ]:
#남은 값은 NaN or 4자리 숫자가 아닌 것
# 남은 값을 -1로 변환 -> astype()메서드로 데이터타입을 정수형인 int32로 변환
ns_book5.loc[unkown_year, '발행년도'] = '-1'
ns_book5 = ns_book5.astype({'발행년도':'int32'})

- **gt() 메서드** : 전달된 값보다 큰 값을 찾음

In [ ]:
#연도가 4000년이 넘는 행의 전체 개수 확인을 위해 sum()함수 함께 사용
ns_book5['발행년도'].gt(4000).sum()

np.int64(131)

**< gt()메서드 대신 () 기호를 쓸 수 있음 >** \
ns_book5['발행년도'].gt(4000).sum() \
= (ns_book5['발행년도'] > 4000).sum() \

데이터 프레임 원소 비교할 수 있는 메서드\
gt() : > : 지정된 값보다 큰 값 검사\
ge() : >= : 지정된 값보다 크거나 같은 값 검사\
It() : < : 지정된 값보다 작은 값을 검사\
le() : <= : 지정된 값보다 작거나 같은 값을 검사\
eq() : == : 지정된 값과 같은 값을 검사\
ne() : != : 지정된 값과 같지 않은 값을 검사

In [ ]:
#4000년이 넘는 연도에서 2333을 빼서 서기로 바꾼 다음 4,000년이 넘는 도서가 있는지 확인

# 4000보다 큰 값을 찾기
dangun_yy_rows = ns_book5['발행년도'].gt(4000)
# 찾은 값에서 2333을 빼서 서기로 바꿈
ns_book5.loc[dangun_yy_rows, '발행년도'] = ns_book5.loc[dangun_yy_rows, '발행년도'] - 2333
# 다시 4000보다 큰 값을 찾음
dangun_year = ns_book5['발행년도'].gt(4000)
print(dangun_year.sum())
ns_book5[dangun_year].head(2)

13


번호                                         도서명               저자  \
222858  234268                                  Modern art  [computer file]   
270269  282852  현대 영어학=Linguistic theory in modern english              이재영   

            출판사  발행년도           ISBN 세트 ISBN 부가기호    권   주제분류번호  도서권수  대출건수  \
222858  GROLIER  7611  9780717233243          NaN  NaN  609.205     1     1   
270269     열림기획  7634  9788986705072            1  NaN    740.1     1     6   

              등록일자  
222858  2009-05-07  
270269  2007-04-14

In [ ]:
#연도가 높은 도서 13권 모두-1로 표시
ns_book5.loc[dangun_year, '발행년도'] = -1

In [ ]:
#연도가 작은 값 확인
#0년도 이상 1900년도 이하 찾기

old_books = ns_book5['발행년도'].gt(0) & ns_book5['발행년도'].lt(1900)
ns_book5[old_books]

번호                                           도서명  \
61149    64093         고흐 씨, 시 읽어 줄까요 :내 마음을 알아주는 시와 그림의 만남    
70566    74347                                  정신 병리학의 문제들    
79550    84164  (최근 7년간) 중요 민법판례 :[2009년 1월~2016년 6월 중요 판례]    
147950  157759              (한·중·일) 밥상 문화 :대표음식으로 본 3국 문화비교    
194576  205407                                책으로 뒤덮인 집의 비밀    
287252  300283                                     (밝혀질)우리歷史   

                                   저자      출판사  발행년도           ISBN  \
61149                          이운진 지음      사계절   176  9788958284277   
70566        지은이: 지그문트 프로이트 ;옮긴이: 황보석     열린책들   151  9788932905181   
79550                          이광섭 편저      법학사   163  9788962898651   
147950                         김경은 지음      이가서   132  9788958643012   
194576  N.E. 보드 지음 ;피터 퍼거슨 그림 ;김지현 옮김     개암나무  1015  9788992844413   
287252                          吳在城 著  黎民族史硏究會  1607  9788986892130   

              세트 ISBN 부가기호    권  주제분류번호  도서권수  대출건수        등록일자  
61149                    0  NaN   811.7     1     6  2017-10-27  
70566   9788932905082    9   10   185.5     1     4  2017-04-26  
79550                    9  NaN     365     1    14  2016-09-27  
147950                   0  NaN  381.75     1    30  2013-02-19  
194576                   7  NaN     843     1    15  2010-08-18  
287252                   0  NaN     911     1     5  2006-07-06

In [ ]:
# 잘못된 값을 -1로 설정 후 전체 행 개수 확인
ns_book5.loc[old_books, '발행년도'] = -1
#-1과 같은 값이 몇 개인지 세기
ns_book5['발행년도'].eq(-1).sum()

np.int64(86)

##**7. 누락된 정보 채우기**

In [ ]:
#도서명, 저자, 출판사 열에 누락된 값 or 발행년도 열이 -1인 행 개수 확인
# '\' : 아랫줄에 있는 코드와 이어짐
na_rows = ns_book5['도서명'].isna() | ns_book5['저자'].isna()\
| ns_book5['출판사'].isna() | ns_book5['발행년도'].eq(-1)
print(na_rows.sum())
ns_book5[na_rows].head(2)

5268


번호          도서명          저자  출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
78    79   아산 정주영 레거시         김화진  NaN  2021  9788952129529          NaN  NaN   
265  278  골목의 시간을 그리다  정명섭.김효찬 지음  NaN  2021  9791191266054          NaN  NaN   

    주제분류번호  도서권수  대출건수        등록일자  
78     325     1     1  2021-03-15  
265    NaN     1     0  2021-03-12

In [ ]:
#누락되거나 알 수 없는 행을 뷰티플수프를 사용해 채우기
import requests
from bs4 import BeautifulSoup

In [ ]:
def get_book_title(isbn):
  # URL에 ISBN을 넣어 HTML 가져옵니다.
  url = 'http://www.yes24.com/Product/Search?domain=B00K&query={}'
  r = requests.get(url.format(isbn))
  # HTML 파싱
  soup = BeautifulSoup(r.text, 'html.parser')
  # 클래스 이름이 'gdjame'인 <a> 태그의 텍스트를 가져옵니다.
  title = soup.find('a', attrs={'class': 'gd_name'}).get_text()
  return title

In [ ]:
get_book_title(9791191266054)

'골목의 시간을 그리다'

- **find_all() 메서드** : 저자를 담은 \<a> 태그 모두 추출
- **for문** : \<a> 태그에 속한 모든 텍스트를 파이썬 리스트에 저장
-**join() 메서드** : 하나의 문자열로 합침
- **re모듈, findall() 함수** : 원하는 정규식에 매칭되는 모든 문자열을 찾아 리스트로 반환

In [ ]:
import re

def get_book_info(row):
  title = row['도서명']
  author = row['저자']
  pub= row['출판사']
  year= row['발행년도']
  #Yes24 도서 검색 페이지 URL
  url='http://www.yes24.com/Product/Search?domain=B00K&query={}'
  #URL에 ISBN을 넣어 HTML 가져옵니다
  r = requests.get(url.format(row['ISBN']))
  soup = BeautifulSoup(r.text, 'html.parser') #HTML파싱
  try:
    if pd.isna(title):
      #클래스 이름이 'gd_name'인 <a> 태그의 텍스트를 가져옵니다
      title = soup.find('a', attrs={'class':'gd_name'}).get_txt()
  except AttributeError:
    pass

  try:
    if pd.isna(author):
      #클래스 이름이 'info_auth'인 <span> 태그 아래 <a> 태그의 텍스트를 가져옵니다
      authors = soup.find('span', attrs={'class':'info_auth'}).find_all('a')
      author_list = [auth.get_text() for auth in authors]
      author = ','.join(author_list)
  except AttributeError:
    pass

  try:
    if pd.isna(pub):
    #클래스 이름이 'info_pub'인 <span> 태그 아래 <a> 태그의 텍스트를 가져옵니다
      pub = soup.find('span', attrs={'class':'info_pub'}).find('a').get_text()
  except ArithmeticError:
    pass

  try:
    if year==-1:
      #클래스 이름이 'info_date'인 <span> 태그 아래 텍스트를 가져옵니다.
      year_str = soup.find('span', attrs={'class':'info_date'}).get_text()
      #정규 표현식으로 찾은 값 중에 첫 번째 값만 사용한다.
      year = re.findall(r'\d{4}', year_str)[0]
  except AttributeError:
    pass

  return title, author, pub, year

- 뷰티플수프로 안될 경우 오류 발생\
\-> 다음 요소 추출하도록 **try ~ except문**으로 예외 처리


- **result_type 매개변수**로 반환된 값을 각기 다른 열로 만듦


In [ ]:
updated_sample = ns_book5[na_rows].head(2).apply(get_book_info,\
                                                 axis=1, result_type='expand')
updated_sample

0           1           2     3
78    아산 정주영 레거시         김화진  서울대학교출판문화원  2021
265  골목의 시간을 그리다  정명섭.김효찬 지음      초록비책공방  2021

In [ ]:
gdown.download('https://bit.ly/3UJZiHw', 'ns_book5_update.csv',\
               quiet=False)
ns_book5_update = pd.read_csv('ns_book5_update.csv', index_col=0)
ns_book5_update.head()

Downloading...
From: https://bit.ly/3UJZiHw
To: /content/ns_book5_update.csv
100%|██████████| 406k/406k [00:00<00:00, 102MB/s]


도서명               저자         출판사  발행년도
78                 아산 정주영 레거시              김화진  서울대학교출판문화원  2021
265               골목의 시간을 그리다       정명섭.김효찬 지음      초록비책공방  2021
354                    한국인의 맛           정명섭 지음         추수밭  2021
539              한성부, 달 밝은 밤에           김이삭 지음      고즈넉이엔티  2021
607  100일 완성 마그마 러시아어 중고급 단어장  러포자 구제 연구소 외 지음         문예림  2021

In [ ]:
ns_book5.update(ns_book5_update)
na_rows = ns_book5['도서명'].isna()| ns_book5['저자'].isna()\
 |ns_book5['출판사'].isna()| ns_book5['발행년도'].eq(-1)
print(na_rows.sum())

4615


<ipython-input-42-29a3248acd46>:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[2021 2021 2021 ... 1996 1996 1995]' has dtype incompatible with int32, please explicitly cast to a compatible dtype first.
  ns_book5.update(ns_book5_update)


- **dropna() 메서드**에 열을 리스트로 지정 -\> 누락된 값이 있는 행을 삭제 -\> '발행년도' 열 값이 -1이 아닌 행만 선택하여 ns_book6 데이터프레임 생성

In [ ]:
ns_book5 = ns_book5.astype({'발행년도':'int32'})
ns_book6 = ns_book5.dropna(subset=['도서명', '저자', '출판사'])
ns_book6 = ns_book6[ns_book6['발행년도']!=-1]
ns_book6. head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0          NaN  NaN    NaN     1     0  2021-03-19  
1          NaN  NaN    NaN     1     0  2021-03-19  
2          NaN  NaN    NaN     1     0  2021-03-19  
3          NaN  NaN    NaN     1     0  2021-03-19  
4          NaN  NaN    NaN     1     0  2021-03-19

In [ ]:
ns_book6.to_csv('ns_book6.csv', index=False)

##**8. 데이터를 이애하고 올바르게 정제하기**

In [ ]:
def data_fixing(ns_book4):

#잘못된 값을 수정하거나 NaN을 채우는 함수
# :param ns_book4: data一cleaning。함수에서 전처리된 데이터프레임

# 도서권수와 대출건수를 int32로 바꿉니다.
  ns_book4 = ns_book4.astype({'도서권수':'int32', '대출건수': 'int32'})
 # NaN인 세트 ISBN을 빈 문자열로 바꿉니다.
  set_isbn_na_rows = ns_book4['세트 ISBN'].isna()
  ns_book4.loc[set_isbn_na_rows, '세트 ISBN'] = ''
 # 발행년도 열에서 연도 네 자리를 추출하여 대체합니다. 나머지 발행년도는 -1 로 바꿉니다.
  ns_book5 = ns_book4.replace({'발행년도':'.*(\d{4}).*'}, r'\1', regex=True)
  unkown_year = ns_book5['발행년도'].str.contains('\D', na=True)
  ns_book5.loc[unkown_year, '발행년도'] = '-1'
 # 발행년도를 int32로 바꿉니다.
  ns_book5 = ns_book5.astype({'발행년도': 'int32'})
 # 4000년 이상인 경우 2333년을 뺍니다.
  dangun_yy_rows = ns_book5['발행년도'] .gt(4000)
  ns_book5.loc[dangun_yy_rows, '발행년도'] = ns_book5.loc[dangun_yy_rows,\
                                                      '발행년도'] - 2333
# 여전히 4000년 이상인 경우 -1 로 바꿉니다.
  dangun_year = ns_book5 ['발행년도'].gt (4000)
  ns_book5.loc[dangun_year, '발행년도'] = -1
 # 0-1900^ 사이의 발행년도는 -1 로 바꿉니다.
  old_books = ns_book5['발행년도'].gt(0) & ns_book5['발행년도'].lt(1900)
  ns_book5.loc[old_books, '발행년도'] = -1
 # 도서명, 저자, 출판사가 NaN이거나 발행년도가 -1 인 행을 찾습니다.
  na_rows = ns_book5['도서명'].isna() | ns_book5['저자'].isna() \
  | ns_book5['출판사'].isna() | ns_book5['발행년도'].eq(-1)
 # Yes24 도서 상세 페이지에서 누락된 정보를 채웁니다.
  updated_sample = ns_book5[na_rows].apply(get_book_info,
                                           axis=1, result_type = 'expand')
  # 아래 줄의 들여쓰기를 수정했습니다.
  updated_sample.columns = ['도서명','저자','출판사','발행년도']
  ns_book5.update(updated_sample)
 # 도서명, 저자, 출판사가 NaN이거나 발행년도가 -1 인 행을 삭제합니다.
  ns_book6 = ns_book5.dropna(subset=['도서명','저자','출판사'])
  ns_book6 = ns_book6[ns_book6['발행년도'] != -1]

  return ns_book6

< 핵심 포인트 > \
- **NaN** : 누락된 값 표시하는 기호, \
  isna()메서드 사용 -\> NaN 여부 확인,\
  notna()메서드 사용 -\> NaN이 아닌 값 체크
-**정규 표현식** : 문자열에서 패턴을 찾고 대체하기 위한 규칙의 모음

< 핵심 함수와 메서드 > \

DataFrame.info() : 데이터프레임의 요약 정보 출력\
DataFrame.isna() : 누락된 값을 감지하는 메서드로 셀의 값이 None이나 NaN일 경우 True를 반환\
DataFrame.astype() : 데이터 타입 지정\
DataFrame.fillna() : 데이터프레임에서 누락된 원소의 값을 채움\
DataFrame.replace() : 데이터프레임의 값을 다른 값으로 바꿈\
Series.str.contains() : 시리즈나 인덱스에서 문자열 패턴을 포함하고 있는지 검사\
DataFrame.gt() : 데이터프레임의 원소보다 큰 값을 검사\